tests/task.py

```diff
@@ -0,0 +1,123 @@
+
+from .model import LinearMLP
+from .dataset import XORDataset, ExtraXORDataset
+
+
+def use_cuda():
+    if int(os_getenv('CPU_ONLY', '0')):
+        return False
+    else:
+        return torch.cuda.is_available()
+
+
+class TaskTemplate(type):
+    def __call__(cls, *args, **kwargs):
+        obj = type.__call__(cls, *args, **kwargs)
+        if hasattr(obj, '__post_init__'):
+            obj.__post_init__()
+        return obj
+
+
+def use_metaclass(meta_cls):
+    """ A decorator to make metaclass work on both Py2k and Py3k. """
+    def wrapper(cls):
+        attrs = vars(cls).copy()
+        attrs.pop('__dict__', None)
+        attrs.pop('__weakref__', None)
+        return meta_cls(cls.__name__, cls.__bases__, attrs)
+    return wrapper
+
+
+@use_metaclass(TaskTemplate)
+class BaseTask(object):
+    def __init__(self):
+        self.batch_size = -1
+        self.model = None
+        self.optimizer = None
+        self.criterion = None
+        self.device = None
+        self.train_loader = None
+        self.val_loader = None
+
+    def __post_init__(self):
+        # check whether cuda is available or not
+        if not use_cuda():
+            self.device = None
+        if self.device is None:
+            return
+
+        if isinstance(self.device, str):
+            try:
+                self.device = torch.device(self.device)
+            except RuntimeError as ex:
+                raise ex
+        elif not isinstance(self.device, torch.device):
+            raise TypeError('Invalid type of device.')
+
+
+class XORTask(BaseTask):
+    def __init__(self, validate=False):
+        super(XORTask, self).__init__()
+        bs, steps = 8, 64
+        dataset = XORDataset(bs*steps)
+        if validate:
+            self.train_loader = DataLoader(Subset(dataset, range(steps-bs)))
+            self.val_loader = DataLoader(Subset(dataset, range(steps-bs, steps)))
+        else:
+            self.train_loader = DataLoader(dataset)
+            self.val_loader = None
+
+        self.batch_size = bs
+        self.model = LinearMLP([8, 4, 1])
+        self.optimizer = optim.SGD(self.model.parameters(), lr=1e-3)
+        self.criterion = nn.MSELoss()
+        self.device = torch.device('cuda')
+
+
+class ExtraXORTask(BaseTask):
+    def __init__(self, validate=False):
+        super(ExtraXORTask, self).__init__()
+        bs, steps = 8, 64
+        dataset = ExtraXORDataset(bs*steps, extra_dims=2)
+        if validate:
+            self.train_loader = DataLoader(Subset(dataset, range(steps-bs)))
+            self.val_loader = DataLoader(Subset(dataset, range(steps-bs, steps)))
+        else:
+            self.train_loader = DataLoader(dataset)
+            self.val_loader = None
+
+        self.model = LinearMLP([8, 4, 1])
+        self.optimizer = optim.SGD(self.model.parameters(), lr=1e-3)
+        self.criterion = nn.MSELoss()
+        self.device = torch.device('cuda')
+
+
+class DifferentLearningRateTask(BaseTask):
```

Comment:
I suggest using `Differential` or `Discriminative` instead of `Different`. Differential is the popular name for this but discriminative is the `official` name; so either one is fine

Reply:
Good point, I was a bit confused while trying to name this task.
It seems `Discriminative` is a good choice for me after seeing [this comment](https://forums.fast.ai/t/what-happened-to-differential-learning-rates-and-warmup/28809/3).